README.md
# ‼️ **Better use [Telethon](https://lonamiwebs.github.io/Telethon/)**

It does have docs


# api-mtproto

[![npm version][npm-image]][npm-url]

**Telegram Mobile Protocol** [(MTProto)](https://core.telegram.org/mtproto) library in **es6**

## About MTProto..

**MTProto** is the [Telegram Messenger](http://www.telegram.org) protocol
_"designed for access to a server API from applications running on mobile devices"_.

The Mobile Protocol is subdivided into three components ([from the official site](https://core.telegram.org/mtproto#general-description)):

 - High-level component (API query language): defines the method whereby API
 queries and responses are converted to binary messages.

 - Cryptographic (authorization) layer: defines the method by which messages
 are encrypted prior to being transmitted through the transport protocol.

 - Transport component: defines the method for the client and the server to transmit
 messages over some other existing network protocol (such as, http, https, tcp, udp).



## api-mtproto in short..

No more additional libs.
The **api-mtproto** library implements the **Mobile Protocol** and provides all features for work with telegram protocol:

 - A high level api for server connection

 - Promise-based API

 - **HTTP connections** implemented in the transport layer

 - **Web worker** support for blazing fast crypto math works in background

 - A cipher implementation for **AES and RSA encryption** in the security layer

 - Both **plain-text and encrypted message** to communicate data with the server

 - **Diffie-Hellman key exchange** supported by the **prime factorization** function implemented in the security layer

 - **MTProto TL-Schema** compilation as **javascript classes and functions**

 - Custom **async storage** support for saving user data between sessions


## Installation

```bash
$ npm install --save api-mtproto@beta
```

## Usage

```javascript
import MTProto from 'api-mtproto'

const phone = {
  num : '+9996620001',
  code: '22222'
}

const api = {
  layer          : 57,
  initConnection : 0x69796de9,
  api_id         : 49631
}

const server = {
  dev: true //We will connect to the test server.
}           //Any empty configurations fields can just not be specified

const client = MTProto({ server, api })

async function connect(){
  const { phone_code_hash } = await client('auth.sendCode', {
    phone_number  : phone.num,
    current_number: false,
    api_id        : 49631,
    api_hash      : 'fb050b8f6771e15bfda5df2409931569'
  })
  const { user } = await client('auth.signIn', {
    phone_number   : phone.num,
    phone_code_hash: phone_code_hash,
    phone_code     : phone.code
  })

  console.log('signed as ', user)
}

connect()
```

Above we used two functions from the API.
```typescript
type auth.sendCode = (phone_number: string, sms_type: int,
  api_id: int, api_hash: string, lang_code: string) => {
    phone_registered: boolean,
    phone_code_hash: string,
    send_call_timeout: int,
    is_password: boolean
  }

type auth.signIn = (phone_number: string, phone_code_hash: string, phone_code: string) => {
  expires: int,
  user: User
}
```
[More][send-code] about [them][sign-in], as well as about many other methods, you can read in the [official documentation][docs].

Additional examples can be obtained from [examples][examples] folder.

## Storage

You can use your own storages like [localForage][localForage] for saving data.
Module accepts the following interface

```typescript
interface AsyncStorage {
  get(key: string): Promise<any>;
  set(key: string, value: any): Promise<void>;
  remove(...keys: string[]): Promise<void>;
  clear(): Promise<void>;
}
```

```javascript
import { MTProto } from 'api-mtproto'
import { api } from './config'
import CustomStorage from './storage'

const client = MTProto({
  api,
  app: {
    storage: CustomStorage
  }
})

```

## License

The project is released under the [Mit License](./LICENSE)

[examples]: https://github.com/zerobias/api-mtproto/tree/develop/examples
[localForage]: https://github.com/localForage/localForage
[docs]: https://core.telegram.org/
[send-code]: https://core.telegram.org/method/auth.sendCode
[sign-in]: https://core.telegram.org/method/auth.signIn
[npm-url]: https://www.npmjs.org/package/api-mtproto
[npm-image]: https://badge.fury.io/js/api-mtproto.svg


CHANGELOG.md
# 2.2.2
* Fixed bug with cyrillic text encoding
* Fixed bug with several broken methods invocations
* Optimized authorization performance
* Added fix for automatic base data center selection by [@goodmind][]

# 2.2.1
* Flag (optional type) fields now acts like common fields.
* Zeroes, empty strings and empty types now can be omited. write only useful fields.
* *invokeWithLayer* api field now may detects internally and don't required (but still valid).
* Type check argument fields
* Fix auth race condition
* Add batch async logger

# 2.2.0

* **breaking** Instance now creates without `new`
* **breaking** Rename module exports from `ApiManager` to `MTProto`

# =<2.1.0

Several early alpha versions based on new architechture

---

# 1.1.0 *(beta)*

* **breaking** Remove all functions from response. Just use the field values.
* Remove logger from response
* Add changelog.md

# 1.0.6 *(beta)*

* Https connection. Usage:
```javascript
const { network } = require('api-mtproto')
const connection = network.http({ host: 'ip', port: '80', protocol: 'https' })
```
* Websockets connection. Usage:
```javascript
const connection = network.wc({ host: 'ip', port: '80' })
```
* Precision timing
* Major performance boost

[@goodmind]: https://github.com/goodmind/

index.d.ts
declare module 'api-mtproto' {
  type DC = {
    id: number
    host: string
    port: number
  }
  type ServerConfig = {
    dev?: boolean
    webogram?: boolean
    dcList?: Array<DC>
  }
  type PublicKey = {
    modulus: string
    exponent: string
  }
  type AppConfig = {
    debug?: boolean
    publicKeys?: Array<PublicKey>
    storage?: AsyncStorage
  }
  type ApiConfig = {
    invokeWithLayer?: number
    layer          ?: number
    initConnection ?: number
    api_id         ?: number
    device_model   ?: string
    system_version ?: string
    app_version    ?: string
    lang_code      ?: string
  }
  type Config = {
    server?: ServerConfig
    api?: ApiConfig
    app?: AppConfig
    schema?: Object
    mtSchema?: Object
  }

  export interface ApiManagerInstance {
    readonly storage: AsyncStorage
    readonly updates: any
    <T>(method: string): Promise<T>
    <T>(method: string, params: Object): Promise<T>
    <T>(method: string, params: Object, options: Object): Promise<T>
    setUserAuth<T>(dc: number, userAuth: T): void
    on(event: string|string[], handler: Function): void
  }
  interface IApiManager {
    new (): ApiManagerInstance
    new ({ server, api, app, schema, mtSchema }: Config): ApiManagerInstance
  }
  export const ApiManager: IApiManager
  class ApiManagerClass {
    readonly storage: AsyncStorage
    setUserAuth<T>(dc: number, userAuth: T): void
    on(event: string|string[], handler: Function): void
  }
  export interface AsyncStorage {
    get(key: string): Promise<any>
    set(key: string, val: any): Promise<any>
    remove(...keys: string[]): Promise<any>
    clear(): Promise<{}>
  }

  function MTProto({ server, api, app, schema, mtSchema }: Config): ApiManagerInstance
  export default MTProto
}


src/util/log.js
//@flow

// import memoize from 'memoizee'

import Debug from 'debug'

import trim from 'ramda/src/trim'
import map from 'ramda/src/map'
import chain from 'ramda/src/chain'
import pipe from 'ramda/src/pipe'
import split from 'ramda/src/split'
import both from 'ramda/src/both'
import is from 'ramda/src/is'
import when from 'ramda/src/when'
import take from 'ramda/src/take'
import reject from 'ramda/src/reject'
import isEmpty from 'ramda/src/isEmpty'
import join from 'ramda/src/join'
import unapply from 'ramda/src/unapply'
import unnest from 'ramda/src/unnest'
import tail from 'ramda/src/tail'

import dTime from './dtime'
import { immediate } from './smart-timeout'

type VariString = string | string[]

type Normalize = (tag: string) => string
type FullNormalize = (tags: VariString) => string

const tagNormalize: Normalize = e => `[${e}]`

const arrify = unapply(unnest)

const fullNormalize: FullNormalize = pipe(
  arrify,
  chain(split(',')),
  map(trim),
  reject(isEmpty),
  map(tagNormalize),
  join('')
)

const stringNormalize = when(
  both(is(String), e => e.length > 50),
  take(150)
)
// const isSimple = either(
//   is(String),
//   is(Number)
// )

// const prettify = unless(
//   isSimple,
//   pretty
// )

const genericLogger = Debug('api-mtproto')

class LogEvent {
  log: typeof genericLogger
  values: mixed[]
  constructor(log: typeof genericLogger, values: mixed[]) {
    this.log = log
    this.values = values
  }
  print() {
    this.log(...this.values)
  }
}

class Sheduler {
  queue: LogEvent[][] = []
  buffer: LogEvent[] = []
  add = (log: typeof genericLogger, time: string, tagStr: string, values: mixed[]) => {
    const results = values.map(stringNormalize)
    const first = results[0] || ''
    const other = tail(results)
    const firstLine = [tagStr, time, first].join('  ')
    this.buffer.push(new LogEvent(log, [firstLine, ...other]))
  }
  sheduleBuffer = () => {
    this.queue.push(this.buffer)
    this.buffer = []
  }
  print = () => {
    for (const buffer of this.queue)
      for (const logEvent of buffer)
        logEvent.print()
    this.queue = []
  }
  constructor() {
    setInterval(this.sheduleBuffer, 50)
    setInterval(this.print, 300)
  }
}

const sheduler = new Sheduler

const Logger = (moduleName: VariString, ...rest: string[]) => {
  const fullModule: string[] = arrify(moduleName, ...rest)
  fullModule.unshift('api-mtproto')
  const fullname = fullModule.join(':')
  const debug = Debug(fullname)
  const logger = (tags: string | string[]) => {
    const tagStr = fullNormalize(tags)
    return (...objects: any[]) => {
      const time = dTime()
      immediate(sheduler.add, debug, time, tagStr, objects)
    }
  }
  return logger
}

export const setLogger = (customLogger: Function) => {
  Debug.log = customLogger
}

export default Logger

examples/node-storage.js
const { MTProto } = require('api-mtproto')
const { Storage } = require('mtproto-storage-fs')
const readline = require('readline')

// The api_id and api_hash values can be obtained here: https://my.telegram.org/
const config = {
  "phone_number": "+1xxxxxxxxx",
  "api_id": 00000,
  "api_hash": "xxxxxx"
}

const app = {
  storage: new Storage('./storage.json')
}

const phone = {
  num: config.phone_number
}

const api = {
  layer         : 57,
  initConnection: 0x69796de9,
  api_id        : config.api_id
}

const server = {
  dev: false
}

const client  = MTProto({ server, api, app })

// This function will stop execution of the program until you enter the code
// that is sent via SMS or Telegram.
const askForCode = () => {
  return new Promise((resolve) => {
    const rl = readline.createInterface({
      input : process.stdin,
      output: process.stdout
    })

    rl.question('Please enter passcode for ' + phone.num + ':\n', (num) => {
      rl.close()
      resolve(num)
    })
  })
}

// First you will receive a code via SMS or Telegram, which you have to enter
// directly in the command line. If you entered the correct code, you will be
// logged in and the credentials are saved.
const login = async (client, phone) => {
  const { phone_code_hash } = await client('auth.sendCode', {
    phone_number  : phone.num,
    current_number: false,
    api_id        : config.api_id,
    api_hash      : config.api_hash
  })

  const phone_code = await askForCode()
  console.log(`Your code: ${phone_code}`)

  const { user } = await client('auth.signIn', {
    phone_number   : phone.num,
    phone_code_hash: phone_code_hash,
    phone_code     : phone_code
  })

  console.log('signed as ', user)
}

const getDialogs = async () => {
  const dialogs = await client('messages.getDialogs', {
    limit: 100,
  })
  console.log('dialogs', dialogs)
}

// First check if we are already signed in (if credentials are stored). If we
// are logged in, execution continues, otherwise the login process begins.
(async function() {
	if (!(await app.storage.get('signedin'))) {
    console.log('not signed in')

		await login(client, phone).catch(console.error)

    console.log('signed in successfully')
		app.storage.set('signedin', true)
	} else {
    console.log('already signed in')
  }
  getDialogs()
})()